tendermint/src/serializers.rs

```diff
@@ -96,24 +97,83 @@ pub mod time_duration {
 pub mod bytes {
+    use serde::{
+        de::{Error, Visitor},
+        Deserialize, Deserializer,
+    };
+    use std::fmt;
+    use subtle_encoding::{base64, hex};
+
+    /// ByteStringType defines the options what an incoming string can represent.
+    enum ByteStringType {
+        Hex,
+        Base64,
+        Regular,
+    }
+
+    /// The Visitor struct to decode the incoming string.
+    struct BytesVisitor {
+        string_type: ByteStringType,
+    }
+
+    /// The Visitor implementation
+    impl<'de> Visitor<'de> for BytesVisitor {
+        type Value = Vec<u8>;
+
+        /// Description of expected input
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            match self.string_type {
+                ByteStringType::Hex => {
+                    formatter.write_str("Hex-encoded byte-array in a String or null")
+                }
+                ByteStringType::Base64 => {
+                    formatter.write_str("Base64-encoded byte-array in a String or null")
+                }
+                ByteStringType::Regular => formatter.write_str("Byte-array in a String or null"),
+            }
+        }
+
+        /// If incoming is 'null', return an empty array.
+        fn visit_none<E>(self) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            Ok(vec![])
+        }
+
+        /// Decode the incoming string based on what string type it is.
+        fn visit_some<D>(
+            self,
+            deserializer: D,
+        ) -> Result<Self::Value, <D as Deserializer<'de>>::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let string = String::deserialize(deserializer)?;
+
+            match self.string_type {
+                ByteStringType::Hex => hex::decode_upper(&string)
+                    .or_else(|_| hex::decode(&string))
+                    .map_err(Error::custom),
+                ByteStringType::Base64 => base64::decode(&string).map_err(Error::custom),
+                ByteStringType::Regular => Ok(string.as_bytes().to_vec()),
+            }
+        }
+    }
 
     /// Serialize into hexstring, deserialize from hexstring
     pub mod hexstring {
-        use serde::{
-            de::Error as deError, ser::Error as serError, Deserialize, Deserializer, Serializer,
-        };
+        use serde::{ser::Error, Deserializer, Serializer};
         use subtle_encoding::hex;
 
         /// Deserialize hexstring into Vec<u8>
         pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
         where
             D: Deserializer<'de>,
         {
-            let string = String::deserialize(deserializer)?;
-
-            hex::decode_upper(&string)
-                .or_else(|_| hex::decode(&string))
-                .map_err(deError::custom)
+            deserializer.deserialize_option(super::BytesVisitor {
+                string_type: super::ByteStringType::Hex,
+            })
```

Comment:
So this `deserialize_option` here and for the "ByteTypes" introduced the need for a Visitor, right? 

No need to apply this but if replaced by sth along the lines of:

```suggestion
let string = Option::<String>::deserialize(deserializer)?.unwrap_or("".into());
            hex::decode_upper(&string)
                .or_else(|_| hex::decode(&string))
                .map_err(serde::de::Error::custom)
```
here and in the other two places and if I'm not wrong the need for a custom Visitor would be gone?

Reply:
Confirmed: https://github.com/informalsystems/tendermint-rs/pull/271/files